### PR TITLE
CXX: Fix bux #1134

### PR DESCRIPTION
**Pull request by pragmaware**
Implement skipping of parenthesis after ignored tokens specified as -I SOMETHING+.
Fixes bug #1134

### CODE REVIEW

File: parsers/cxx/cxx_parser_tokenizer.c [hunk 1]
```diff
@@ -1088,16 +1140,30 @@ boolean cxxParserParseNextToken(void)
...
 			{
 				CXX_DEBUG_PRINT("Ignore token %s",vStringValue(t->pszWord));
-				// FIXME: Handle ignore parens!
+
 				if(szReplacement && *szReplacement)
 				{
+					CXX_DEBUG_PRINT("The token has replacement %s: applying",szReplacement);
 					vStringClear(t->pszWord);
 					vStringCatS(t->pszWord,szReplacement);
+					// FIXME: Do we need to interpret the replacement?
```
Comment (b4n) [inline parsers/cxx/cxx_parser_tokenizer.c]: it probably would be great, but I also think it's not likely to be very important.  But yeah if it could understand replacements to at least keywords like `const`/`extern`/`static` and combination of those, it probably would be awesome for when ignore tokens are really useful.

Comment (pragmaware) [inline parsers/cxx/cxx_parser_tokenizer.c]: I can easily handle single keywords. Combinations are harder to do since they would require the token to be splitted again with a custom algorithm, then one would need to keep a backlog of tokens yet to be returned which would affect the performance of the existing algo... I don't think it's worth the overhead. I'll add a check for keywords.

File: parsers/cxx/cxx_parser_tokenizer.c [hunk 2]
```diff
@@ -1161,7 +1161,45 @@ boolean cxxParserParseNextToken(void)
...
 
 				if(szReplacement && *szReplacement)
-					return TRUE; // already have a token to return
+				{
+					// Already have a token to return.
+					// Check again for keywords.
+					
+					// There is a bit of code duplication here but I'd rather
+					// avoid a function call and a switch as the keyword check
+					// must be fast (especially the one above).
```
Comment (b4n) [inline parsers/cxx/cxx_parser_tokenizer.c]: would it be so slow or problematic, really?  at worse, an inlining hint to the compiler should help.
I'm a bit afraid of the code duplication here because it's already fairly non-trivial, so likely to grow even more complex in the future.

If the part handling `__attribute__` differently make it too annoying/complex, maybe just leave this part out, and simply have a `ccxParserLookupKeyword()` that only does the lookup and special case handling.

Comment (masatake) [inline parsers/cxx/cxx_parser_tokenizer.c]: I support @b4n's comment.

Comment (pragmaware) [inline parsers/cxx/cxx_parser_tokenizer.c]: There :)

File: parsers/cxx/cxx_parser_tokenizer.c [hunk 3]
```diff
@@ -1165,43 +1169,10 @@ boolean cxxParserParseNextToken(void)
...
-							if(iCXXKeyword == CXXKeyword__ATTRIBUTE__)
-							{
-								// special handling for __attribute__
-								return cxxParserParseNextTokenCondenseAttribute();
-							}
-						}
-					}
-
-					return TRUE;
+					goto check_keyword;
```
Comment (b4n) [inline parsers/cxx/cxx_parser_tokenizer.c]: while tempting, this can lead to infinite loop with inappropriate ignore tokens referencing one another, e.g. `-Ifoo=foo`

Comment (masatake) [inline parsers/cxx/cxx_parser_tokenizer.c]: Should I add code for avoiding circular referencing to the main part?

Comment (pragmaware) [inline parsers/cxx/cxx_parser_tokenizer.c]: I guess it's hard as loops can have an arbitrary number of steps.. I'll better just add a simple check in the cxx parser to disallow recursive ignore tokens.

**Comment by masatake** [comment]
I would like to review this.Please,give me time.

**Comment by masatake** [comment]
LGTM. Sorry for being late.

**Comment by pragmaware** [comment]
My last conflict fixing commit seems to be quite big. See a734947. However the cumulative changes listed in this page seem to be OK. Am I messing up something?

**Comment by masatake** [comment]
I think it is better to do `git rebase master` after updating master.

**Comment by masatake** [comment]
If github interface convinces you, you can do everything locally and the do `git push upstream master:master`. A bit dangerous but sometimes I do so.

`git push --force upstream master:master` is forbidden in github side. So we will not get serious trouble through the direct accessing to `master` branch of upstream.

File: parsers/cxx/cxx_parser_tokenizer.c (hunk 3 above)
Comment (b4n) [inline parsers/cxx/cxx_parser_tokenizer.c]: Yeah resolving loops in the main part would be very tricky, and what do actually do with them?  Better have the user side not re-consider replacements (or if it has to, drop ignore tokens once processed, C-preprocessor-style).

File: parsers/cxx/cxx_parser_tokenizer.c [hunk 4]
```diff
@@ -1079,29 +1136,58 @@ bool cxxParserParseNextToken(void)
...
+						goto check_keyword;
+					}
+	
 					return cxxParserParseNextToken();
 				}
 			}
 		}
 
-		t->bFollowedBySpace = t->bFollowedBySpace | isspace(g_cxx.iChar);
+		t->bFollowedBySpace = isspace(g_cxx.iChar);
```
Comment (b4n) [inline parsers/cxx/cxx_parser_tokenizer.c]: How is that related?

**Comment by pragmaware** [comment]
I'm a bit confused with the merging and not sure if it's OK anymore. I think it's safer if I re-do a new clean branch, reapply and recreate the pull request.

File: parsers/cxx/cxx_parser_tokenizer.c (hunk 4 above)
Comment (pragmaware) [inline parsers/cxx/cxx_parser_tokenizer.c]: If I'm not wrong that t->bFollowedBySpace should always read to false on that line.

**Comment by masatake** [comment]
Feel free to do 'git push --force' this PR (as I did frequently :-P).

**Comment by pragmaware** [comment]
Recreated as pull #1151
